common/utils/include/claragenomics/utils/cudasort.cuh

```diff
@@ -0,0 +1,216 @@
+                                    temp_storage_bytes, // <- this one gets changed
+                                    unsorted_keys_d,
+                                    sorted_keys_d,
+                                    unsorted_values_d,
+                                    sorted_values_d,
+                                    number_of_elements,
+                                    begin_bit,
+                                    end_bit);
+
+    // allocate temp storage
+    if (temp_storage_bytes > temp_storage_vect_d.size())
+    {
+        temp_storage_vect_d.resize(temp_storage_bytes);
+    }
+    temp_storage_d     = static_cast<void*>(temp_storage_vect_d.data().get());
+    temp_storage_bytes = temp_storage_vect_d.size();
+
+    // perform actual sort
+    cub::DeviceRadixSort::SortPairs(temp_storage_d,
+                                    temp_storage_bytes,
+                                    unsorted_keys_d,
+                                    sorted_keys_d,
+                                    unsorted_values_d,
+                                    sorted_values_d,
+                                    number_of_elements,
+                                    begin_bit,
+                                    end_bit);
+}
+
+} // namespace details
+
+/// \brief Sorts array by more significant key. Then sorts subarrays belonging to each more significiant key by less significant key
+///
+/// For example:
+/// ms_key ls_key value
+///      6      1     1
+///      2      4     2
+///      5      5     3
+///      4      5     4
+///      4      2     5
+///      2      1     6
+/// yields:
+///      2      1     6
+///      2      4     2
+///      4      2     5
+///      4      5     4
+///      5      5     3
+///      6      1     1
+///
+/// \param more_significant_keys sorted on output
+/// \param less_significant_keys sorted on output
+/// \param values sorted on output
+/// \param max_value_of_more_significant_key optional, defaults to max value for MoreSignificantKeyT
+/// \param max_value_of_less_significant_key optional, defaults to max value for LessSignificantKeyT
+/// \tparam MoreSignificantKeyT
+/// \tparam LessSignificantKeyT
+/// \tparam ValueT
+template <typename MoreSignificantKeyT,
+          typename LessSignificantKeyT,
+          typename ValueT>
+void sort_by_two_keys(thrust::device_vector<MoreSignificantKeyT>& more_significant_keys,
+                      thrust::device_vector<LessSignificantKeyT>& less_significant_keys,
+                      thrust::device_vector<ValueT>& values,
+                      const MoreSignificantKeyT max_value_of_more_significant_key = std::numeric_limits<MoreSignificantKeyT>::max(),
+                      const LessSignificantKeyT max_value_of_less_significant_key = std::numeric_limits<LessSignificantKeyT>::max())
+{
+    // Radix sort is done in-place, meaning sorting by less significant and then more significant keys yields the wanted result
+    //
```

Comment:
Another question: This method assumes that CUB's radix sort is stable, right? Is it?

Reply:
This is a valid point. I just checked https://nvlabs.github.io/cub/structcub_1_1_device_radix_sort.html and they do not mention stability. On the other hand because of the the way CUB is implemented it "has" to be stable. CUB first sorts by a few least significant bits of keys, then the next few "a bit more significant" bits and so on. That's why it "has" to be stable.
I'll check with CUB team if there is actually a guarantee for that

Reply:
Ok, but we should keep in mind that we rely on undocumented behavior here.